presto-main/src/main/java/io/prestosql/server/security/UserExtractionUtil.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server.security;
+
+import java.security.Principal;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Strings.emptyToNull;
+
+final class UserExtractionUtil
+{
+    private UserExtractionUtil() {}
+
+    public static Optional<String> extractUser(Principal principal, Optional<Pattern> extractionPattern)
+    {
+        if (!extractionPattern.isPresent()) {
+            return Optional.empty();
+        }
+
+        Matcher matcher = extractionPattern.get().matcher(principal.getName());
+        if (!matcher.matches()) {
+            return Optional.empty();
+        }
+
+        for (int i = 0; i < matcher.groupCount(); i++) {
+            String group = emptyToNull(matcher.group(i + 1));
+            if (group != null) {
+                return Optional.of(group);
```

Comment:
Do we need "return first non-empty matching group"?

I'd prefer something simpler: "verify pattern has one matching group; return it"

Reply:
This allows a regular expression to contain ORs.  Say for example, you want `(.*)@foo|(.*)@bar`. I know you can put the or in after the `@` sign, but there are more complex cases that would need this.